Doc/regions.md

```diff
@@ -0,0 +1,517 @@
+ It represents any operation that should prevent the region from being accessed by the current thread.]
+
+
+To re-establish the correct local reference count, we calculate the updates to LRC as follows.
+First we calculate the set of local objects, L, that are being made reachable from this region.
+As the set L is only reachable by local objects, we know that its reference count is purely from local objects.  
+Thus, we calculate sum of all the rcs in L, called R.
+Some of these references are internal to the set L,
+and one is from the region itself,
+so we calculate the number of references from L to L, called I.
+Thus, we have created R-I-1 additional references into the region.
+However, L can also reach into the region, so we calculate the number of references from L to the region, C.
+Thus, we need to increase the region's local reference count by R - I - 1, and then subtract C from it.
+This re-establishes the invariant.
+
+To illustrate this, consider the following object graph:
+```mermaid
+graph
+    frame -->|z| obj1
+    frame -->|r| region
+    obj3  -->|f| obj2
+    obj3  -->|g| obj4
+    region-->|f| obj1
+    frame -->|x| obj2
+    frame -->|y| obj3
+    obj2  -->|f| obj1
+    obj2  -->|f| obj4
+    subgraph " "
+        region
+        obj1
+    end
+```
+Now consider executing the following code:
+```python
+z.f = x
+```
+This creates a link from `obj1` to `obj2`.
+The updated object graph is below:
+```mermaid
+graph
+    frame -->|z| obj1
+    frame -->|r| region
+    obj3  -->|f| obj2
+    obj3  -->|g| obj4
+    region-->|f| obj1
+    frame -->|x| obj2
+    frame -->|y| obj3
+    obj1  -->|f| obj2
+    obj2  -->|f| obj1
+    obj2  -->|f| obj4
+    subgraph " "
+        region
+        obj1
+        obj2
+        obj4
+    end
+```
+
+The local reference count for the `region` is modified in the following way.
+* The reference count for `obj2` is increased by 1 to represent the new reference from `obj1`.
+* The objects `obj2` and `obj4` are added to the region, they have a combined reference count of 5.
+* The edge from `obj2` to `obj4` is considered internal, so we reduce the LRC by one.
+* The edge from `obj2` to `obj1` is considered internal, so we reduce the LRC by one.
+* The edge from `obj1` to `obj2` is considered internal, so we reduce the LRC by one.
+
+Thus, the resulting change is to increase the LRC by 5, and then reduce it by 3. 
+
+
+### Immutable objects
+
+To share objects between behaviours, we need to ensure that the object is not mutated.
+This is achieved by making the object "deeply immutable".
+This means that the object and all objects reachable from it are also immutable.
+
+The types in Python are mutable, but are typically only mutated at the start of the program.
+To support this, we need an operation to turn initially mutable objects into immutable objects.
+We extend the programming model with a `freeze` operation that makes an object and its reachable graph of objects immutable.
```

Comment:
Didn't Guido hate this before? Making a local object immutable can potentially propagate to anything local. What if we only support freezing a (sendable) region? It may still cause "surprise immutables", but I think that would be less surprising.

Reply:
Yeah, not sure.  This seemed the most flexible, and didn't seem to introduce any more complexity.  Happy to restrict. I'll add a note.

Reply:
There's not much of a difference between the two options, from CPython's current implementation -- in both cases you basically need to add a "frozen" bit to every object, and honor it in every mutable operation. E.g. if a list is frozen, its `append()` and `__setitem__()` methods (and many others) should raise exceptions, but `__len__()` and `__getitem__()` should keep working. Unfortunately there are many object types that need such checks (e.g. function and class objects).

Reply:
Yeah, this will require a lot of manual reviewing.